lib/style-engine/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,157 @@
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Returns a CSS ruleset based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param string|array  $style_value A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array<string> $path        An array of strings representing a path to the style value.
+	 *
+	 * @return array A CSS ruleset compatible with add_style().
+	 */
+	protected function get_block_style_css_rules( $style_value, $path ) {
+		$style_definition = _wp_array_get( static::BLOCK_STYLE_DEFINITIONS_METADATA, $path, null );
+
+		if ( ! empty( $style_definition ) ) {
+			if (
+				isset( $style_definition['value_func'] ) &&
+				is_callable( $style_definition['value_func'] )
+			) {
+				return call_user_func( $style_definition['value_func'], $style_value, $style_definition['property_key'] );
+			}
+		}
+
+		return array();
+	}
+
+	/**
+	 * Returns an CSS ruleset destined to be inserted in an HTML `style` attribute.
+	 * Styles are bundled based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param array $block_styles An array of styles from a block's attributes.
+	 * @param array $options = array(
+	 *     'inline' => (boolean) Whether to return inline CSS rules.
+	 * );.
+	 *
+	 * @return string A CSS ruleset formatted to be placed in an HTML `style` attribute.
+	 */
+	public function generate( $block_styles, $options = array() ) {
+		$output = '';
+
+		if ( empty( $block_styles ) ) {
+			return $output;
+		}
+
+		$rules = array();
+
+		foreach ( $block_styles as $block_style => $block_style_properties ) {
+			if ( empty( $block_style ) || ! is_array( $block_style_properties ) ) {
+				continue;
+			}
+			$path        = array( $block_style, key( $block_style_properties ) );
+			$style_value = _wp_array_get( $block_styles, $path, null );
+
+			if ( empty( $style_value ) ) {
+				continue;
+			}
+
+			$rules = array_merge( $rules, $this->get_block_style_css_rules( $style_value, $path ) );
+		}
+
+		if ( ! empty( $rules ) ) {
+			// Generate inline style rules.
+			if ( isset( $options['inline'] ) && true === $options['inline'] ) {
+				foreach ( $rules as $rule => $value ) {
+					$output .= "{$rule}:{$value};";
+				}
+			}
+		}
+
+		return $output;
+	}
+}
+
+/**
+ * Returns a CSS ruleset for box model styles such as margins, padding, and borders.
+ *
+ * @param string|array $style_value    A single raw Gutenberg style attributes value for a CSS property.
+ * @param string       $style_property The CSS property for which we're creating a rule.
+ *
+ * @return array The class name for the added style.
+ */
+function gutenberg_get_style_engine_css_box_rules( $style_value, $style_property ) {
```

Comment:
True, thanks for raising. 

I left it out because these formatting functions could really live anywhere and may do perform some specific operations (like gap), so I didn't want to set a precedent for creating too many methods in the class. 

If you think this one in particular should live in the class, I'm all for it!

Reply:
I don't feel too strongly either way, but I suppose if we have it in the class, we have one fewer `gutenberg` prefixes to deal with? We can always rearrange in follow-ups if need be, though 🙂

Reply:
Fewer search and replaces at the start sounds like something I'm down for 😆 

Something like `'value_func'   => 'self::get_css_box_rules',` seems to work as well.